sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml

```diff
@@ -78,14 +96,157 @@ class HasToHex a where
-  toHex = primitive @"BEEncodeHex"
-
-instance HasFromHex Text where
-  fromHex "" = ""
-  fromHex arg | Text.isHex arg = primitive @"BEDecodeHex" (Text.asciiToLower arg)
-  fromHex arg = error $ "Argument is not a hex string: " <> arg
+instance HasFromHex (Optional Text) where
+  fromHex hex | Text.isHex hex = Some(primitive @"BEDecodeHex" (Text.asciiToLower hex))
+  fromHex _ = None
+
+instance HasFromHex (Optional Int) where
+  fromHex "" = Some 0
+  fromHex hex | Text.isHex hex && Text.length hex == 2 = Some(hexByteToInt hex)
+  fromHex hex | Text.isHex hex = do
+    right <- fromHex(Text.substring ((Text.length hex) - 2) 2 hex)
+    left <- fromHex(Text.substring 0 ((Text.length hex) - 2) hex)
+    Some(left * 256 + right)
+  fromHex _ = None
+
+instance HasFromHex (Optional Party) where
+  fromHex hex | Text.isHex hex = do
+    txt <- fromHex hex
+    partyFromText txt
+  fromHex _ = None
+
+-- Numerics can be thought of as a decimal string with a scale. As usage sites need to know the scale, we can
+-- simply hex encode the string representation of the underlying numeric's decimal string. This allows `parseNumeric`
+-- to be used to define `HasFromHex` instances.
+numericViaStringFromHex: NumericScale n => BytesHex -> Optional (Numeric n)
+numericViaStringFromHex hex | Text.isHex hex = do
+  txt <- fromHex hex
+  Text.parseNumeric txt
+numericViaStringFromHex _ = None
+
+-- | Number of bytes present in a byte encoded string.
+byteCount: BytesHex -> Int
+byteCount "" = 0
+byteCount hex | Text.isHex hex = (Text.length hex) / 2
+byteCount arg = error $ "Expected a hex encoded string but argument was: " <> arg
+
+-- | HIDE
+constList: Int -> a -> [a]
+constList 0 _ = []
+constList n x = x :: (constList (n - 1) x)
+
+-- | HIDE
+byteRepeat: Int -> BytesHex -> BytesHex
+byteRepeat len byte | Text.isHex byte && byteCount byte == 1 && len >= 0 = Text.implode(constList len byte)
+byteRepeat _ arg | not (Text.isHex arg) = error $ "Expected a hex encoded string but argument was: " <> arg
+byteRepeat _ hex | Text.isHex hex && byteCount hex /= 1 = error $ "Expected a byte value but hex encoded string was: " <> hex
+byteRepeat len _ = error $ "Expected a positive length but argument was: " <> show len
+
+-- | Minimum Bytes32 hex value
+minBytes32Hex: BytesHex
+minBytes32Hex = byteRepeat 32 "00"
+-- | Maximum Bytes32 hex value
+maxBytes32Hex: BytesHex
+maxBytes32Hex = byteRepeat 32 "ff"
+
+-- | Validate that the byte encoded string is Bytes32Hex
+isBytes32Hex: BytesHex -> Bool
+isBytes32Hex hex | Text.isHex hex = byteCount hex == 32
+isBytes32Hex arg = error $ "Expected a hex encoded string but argument was: " <> arg
+
+-- | Minimum UInt32 hex value
+minUInt32Hex: BytesHex
+minUInt32Hex = byteRepeat 5 "00"
+-- | Maximum UInt32 hex value
+maxUInt32Hex: BytesHex
+maxUInt32Hex = byteRepeat 5 "ff"
+
+-- | Validate that the byte encoded string is UInt32Hex
+isUInt32Hex: BytesHex -> Bool
+isUInt32Hex hex | Text.isHex hex = byteCount hex == 5
```

Comment:
I'm not sure if I'm misunderstanding the logic here - shouldn't Uint32 require only 4 bytes to encode, Uint64 require 8 bytes, and Uint256 require 32?

i.e.
```
ghci> logBase 256 (2 ** 32)
4.0
ghci> logBase 256 (2 ** 64)
8.0
ghci> logBase 256 (2 ** 256)
32.0
```

Reply:
Begger! Thanks for pointing that error out!

Reply:
Fun fact: a slide rule came in handy on this one :)